File: WaltsonJS C/environments/Lexer.c
/**
 * Lexer module created to read through a file
 * and return lexical items as it comes across them.
 * 
 * Created by Walt Loumakis
 */

#include <ctype.h>
#include <stdio.h>
#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include "Lexer.h"
#include "Types.h"
#include "Lexeme.h"

#define null NULL

/**************************************/
/*     Private Method Prototypes      */
/**************************************/

static Lexeme *lexString();
static Lexeme *lexNumber();
static Lexeme *lexVar();

static char readChar();
static void putBack(char ch);
static void skipWhiteSpace();

/**************************************/
/*              Globals               */
/**************************************/

FILE *fp;
int line;

/**************************************/
/*           Public Methods           */
/**************************************/

void initLexer(char *filename) {
	fp = fopen(filename, "r");
	assert(fp != null);
	line = 1;
}

void closeLexer() {
	fclose(fp);
}

Lexeme *lex() {
	skipWhiteSpace();

	char ch = readChar();

	if (ch == EOF) return newLexeme(ENDofINPUT, UNDEFINED, UNDEFINED, UNDEFINED, line);

	switch(ch) {
		case ':': return newLexeme(COLON, UNDEFINED, UNDEFINED, UNDEFINED, line);
		case '[': return newLexeme(OBRACKET, UNDEFINED, UNDEFINED, UNDEFINED, line);
		case ']': return newLexeme(CBRACKET, UNDEFINED, UNDEFINED, UNDEFINED, line);
		case ',': return newLexeme(COMMA, UNDEFINED, UNDEFINED, UNDEFINED, line);
		case '{': return newLexeme(OBRACE, UNDEFINED, UNDEFINED, UNDEFINED, line);
		case '}': return newLexeme(CBRACE, UNDEFINED, UNDEFINED, UNDEFINED, line);
		case '-': return newLexeme(MINUS, UNDEFINED, UNDEFINED, UNDEFINED, line);
		case '\"':
			putBack(ch);
			return lexString();
		default:
			if (isdigit(ch) || ch == '.') {
				putBack(ch);
				return lexNumber();
			}
			else if (isalpha(ch) || ch == '_') {
				putBack(ch);
				return lexVar();
			}
			else {
				char *temp = (char *)malloc(sizeof(char) + 1);
				temp[0] = ch;
				temp[1] = 0;
				return newLexeme(UNKNOWN_CHAR, UNDEFINED, UNDEFINED, temp, line);
			}
	}
}

/**************************************/
/*          Private Methods           */
/**************************************/

static Lexeme *lexString() {
	int ch = -1;
	int last = -1;
	int length = 64;
	char *buffer = (char *)malloc(sizeof(char) * length + 1);
	int index = 0;

	// This gets the first quote
	ch = readChar();
	
	// Despite IBM currently stating that strings cannot have the \" character,
	// I am implementing a method that allows for \" characters for two reasons:
	//		1.	IBM may choose to implement allowing the \" character later on, and
	//		2.	Even if IBM does not choose to do so, the Watson methods called that take strings
	//			as input will return an error if there is a \" character anyway.
	last = ch;
	while ((ch = readChar()) != EOF && (ch != '\"' || (ch == '\"' && last == '\\'))) {
		if (index == length) {
			length *= 2;
			buffer = (char *)realloc(buffer, sizeof(char) * length + 1);
		}
		buffer[index++] = ch;
        buffer[index] = 0;
        last = ch;
	}
	return newLexeme(STRING, UNDEFINED, UNDEFINED, buffer, line);
}

static Lexeme *lexNumber() {
	int ch;
	int length = 32;
	int index = 0;
	int real = 0;
	char *buffer = (char *)malloc(sizeof(char) * length + 1);
	while ((ch = readChar()) != EOF && (isdigit(ch) || ch == '.')) {
		if (index == length)
			return newLexeme(BAD_NUMBER, UNDEFINED, UNDEFINED, UNDEFINED, line);
		if (ch == '.' && real == 1)
			return newLexeme(BAD_NUMBER, UNDEFINED, UNDEFINED, UNDEFINED, line);
		if (ch == '.')
			real = 1;
		buffer[index++] = ch;
		buffer[index] = 0;
	}
	putBack(ch);
	if (real) return newLexeme(REAL, UNDEFINED, atof(buffer), UNDEFINED, line);
	return newLexeme(INTEGER, atoi(buffer), UNDEFINED, UNDEFINED, line);
}

static Lexeme *lexVar() {
	int ch;
	int length = 64;
	char *buffer = (char *)malloc(sizeof(char) * length + 1);
	int index = 0;

	while ((ch = readChar()) != EOF && (isalpha(ch) || ch == '_')) {
		if (index == length) {
			length *= 2;
			buffer = (char *)realloc(buffer, sizeof(char) * length + 1);
		}
		buffer[index++] = ch;
		buffer[index] = 0;
	}
	putBack(ch);
	return newLexeme(ID, UNDEFINED, UNDEFINED, buffer, line);
}

static char readChar() {
	char ch = fgetc(fp);
	if (ch == '\n') ++line;
	return ch;
}

static void putBack(char ch) {
	if (ch == '\n') --line;
	ungetc(ch, fp);
}

static void skipWhiteSpace() {
	char ch;

	while ((ch = readChar()) != EOF && isspace(ch));

	putBack(ch);
}

File: WaltsonJS/lexical/Lexer.js
let fs = require('fs')

export default class Lexer extends Type {

	constructor(filename) {
		this.file = fs.readFileSync(filename, 'utf-8')
		this.position = 0
		this.line = 1
		return this
	}

	static lex() {
		skipWhiteSpace()
		let ch = readChar()
		console.log('position after reading: ', position, '\tfile.length: ', file.length)
		if (position == file.length)
			return new Lexeme(this.END_OF_INPUT, undefined, line, undefined, undefined)
		switch (ch) {
			case '[':
				return new Lexeme(this.OPEN_BRACKET, undefined, line, undefined, undefined)
			case ']':
				return new Lexeme(this.CLOSE_BRACKET, undefined, line, undefined, undefined)
			case '{':
				return new Lexeme(this.OPEN_BRACE, undefined, line, undefined, undefined)
			case '}':
				return new Lexeme(this.CLOSE_BRACE, undefined, line, undefined, undefined)
			case ':':
				return new Lexeme(this.COLON, undefined, line, undefined, undefined)
			case ',':
				return new Lexeme(this.COMMA, undefined, line, undefined, undefined)
			case '-':
				return new Lexeme(this.MINUS, undefined, line, undefined, undefined)
			case '"':
				putBack(ch)
				return lexString()
			case '\d':
				putBack(ch)
				return lexNumber()
			default:
				if (/\w/.test(ch)) {
					putBack(ch)
					return lexVar()
				}
				else {
					throw Error('Error lexing character ' + ch + '!')
				}
		}
	}

	lexString() {
		let ch = readChar()		// Gets first "
		let ret = ""
		while ((ch = readChar()) != '"') {
			ret += ch
		}
		// Throw away the terminating quote by not pushing back

		return new Lexeme(this.STRING, ret, line, undefined, undefined)
	}

	lexNumber() {
		let ch = readChar()
		let real = 0
		let ret = ""
		while (/\d/.test(ch) || ch == '.') {
			if (real && ch == '.')
				throw Error("Error on line " + line + ": number with two decimals!")
			if (ch == '.')
				real = 1
			ret += ch
		}
		putBack(ch)
		return new Lexeme(this.NUMBER, real ? parseFloat(ret) : parseInt(ret), line, undefined, undefined)
	}

	lexVar() {
		let ch = readChar()
		let ret = ""
		let size = 64
		let index = 0
		while (/\w/.test(ch) || ch == '_') {
			if (index == size)
				throw Error("Error on line " + line + ": variable name is too long!")
			ret += ch
			index++
		}
		putBack(ch)
		return new Lexeme(this.ID, ret, line, undefined, undefined)
	}

	skipWhiteSpace() {
		console.log('skipWhiteSpace')
		let ch = readChar()
		while(/\s/.test(ch)) {
			ch = readChar()
		}
		putBack(ch)
	}

	readChar() {
		console.log('readChar')
		let ch = file[position]
		if (ch == '\n')
			line++
		position++
		return ch
	}

	putBack(ch) {
		console.log('putBack')
		if (ch == '\n')
			line--
		position--
	}

}